State: Introduce functionality for WP.org plugin lists

tyxla opened the pull request:
#### Changes proposed in this Pull Request

This PR introduces action creators, reducers, and selectors to mirror the existing `lib/plugins/wporg-data` flux store. The existing code that uses the old flux store will be ported in another PR.

Part of #24180.

#### Testing instructions

* Verify tests pass.
* If you're interested to manually test it, you can try:
  * Dispatching new actions for requesting and receiving data, then observing the reducer behavior
  * Hooking the action creators and selectors in a component and dispatching actions there, then testing selectors return value.

tyxla committed:
State: Implement basic wporg plugins list fetching

tyxla committed:
State: Implement basic wporg plugins list reducer

tyxla committed:
State: Prevent duplicate plugins list fetch requests

tyxla committed:
State: Safeguard plugins fetching selector access

tyxla committed:
State: Add selector to retrieve next plugins list page

tyxla committed:
State: Add action creator to fetch next category page

tyxla committed:
State: Add inline docs for new functionality

tyxla committed:
State: Add reducer tests

tyxla committed:
State: Add selector tests

tyxla committed:
State: Expand README

sarayourfriend commented on client/state/plugins/wporg/actions.js:
```diff
@@ -52,3 +65,144 @@ export function fetchPluginData( pluginSlug ) {
+ * @param {object} pagination Paginatioin data, as retrieved from the API response.
+ * @returns {Function} Action thunk
+ */
+function receivePluginsList( category, page, searchTerm, data, error, pagination = null ) {
+	return ( dispatch ) =>
+		dispatch( {
+			type: PLUGINS_WPORG_LIST_RECEIVE,
+			category,
+			page,
+			searchTerm,
+			data: normalizePluginsList( data ),
+			pagination,
+			error,
+		} );
+}
```
Does this need to be a `thunk`? It could be just a regular action, no?

tyxla replied:
Yep, good point. Updated in 09db4c427071fc90423a5f03e46f381cae25333f

tyxla committed:
Make receivePluginsList a simple action creator

flootr commented on client/state/plugins/wporg/actions.js:
```diff
@@ -52,3 +65,143 @@ export function fetchPluginData( pluginSlug ) {
+
+		// The "Featured" category is managed by WP.com instead of WP.org
+		if ( 'featured' === category ) {
+			wpcom
```
Nit: should we try to stick to use `async await` syntax to avoid deeper levels of nesting? No strong opinion on my side.

tyxla replied:
I actually wanted to go that way, but then reconsidered when I found that I'd also have to refactor `fetchPluginsList`. Don't want to make this PR too large 😉 

It would be a nice improvement for the future, but definitely not a blocker IMHO. Also, I don't feel strongly either way, the current implementation I have suggested looks pretty clear to me.

flootr commented on client/state/plugins/wporg/reducer.js:
```diff
@@ -45,7 +71,58 @@ export function items( state = {}, action ) {
+	const { category, data, searchTerm, type } = action;
+	switch ( type ) {
+		case PLUGINS_WPORG_LIST_RECEIVE:
+			if ( data ) {
```
What about using an early-return `if ( ! data ) { return state; }` in this case to avoid one level of nesting?

tyxla replied:
Good call. Done in 9fad4151ccd482e33e810f22e12fd27430a27eb7

flootr commented on client/state/plugins/wporg/reducer.js:
```diff
@@ -45,7 +71,58 @@ export function items( state = {}, action ) {
+						...state,
+						category: {
+							...state.category,
+							[ category ]: [ ...prevCategoryState, ...data ],
```
I may be wrong but would this be a good fit for a `keyedReducer`?

tyxla replied:
It would be if we were using objects, but here we gather the plugins in arrays. Also, in the case where we're dealing with paginated data, we don't overwrite previous data, but we only append the new data.

I was thinking to build a couple of functions to remove repetition here, but eventually left it like that to make to more clear as to what we're doing and remove unnecessary indirection.

Do you have any further suggestions or do you think we can leave it as-is?

flootr reviewed (commented):
A few comments, let me know what you think. All in all this looks very solid 👍🏻

tyxla committed:
Early bail

tyxla commented:
> A few comments, let me know what you think. All in all this looks very solid 👍🏻

Thanks @flootr! I've addressed all comments and/or left replies - feel free to take another look at it when you have a chance. This one is kind of important, because it's a blocker for the final work on plugins reduxification 😉

flootr reviewed (approved):
LGTM
